File: advent-of-code/2020/Day15.kt
fun play(start: List<Int>, iteration: Int) {
    val indices = values.dropLast(1).asSequence().withIndex().associate { (index, it) -> it to index }.toMutableMap()

    var last = values.last()

    for (i in values.size until iteration) {
        val index = indices[last]
        indices[last] = i - 1
        last = if (index == null) 0 else i - 1 - index
    }

    return last
}

fun main() {
    val values = readLine()!!.splitToSequence(",").map { it.toInt() }.toList()
    val first = play(values, 2020)
    val second = play(values, 30000000)
    // WARN probably set -Xmx3g to extend heap size
    println("$first $second")
}

File: advent-of-code/2020/Day15.jl
function play(start, iteration)
    indices = Dict{Int, Int}()
    for i in 1:length(start) - 1
        indices[start[i]] = i
    end

    last = start[end]

    for i in length(start):iteration - 1
        index = get(indices, last, -1)
        indices[last] = i
        if index == -1
            last = 0
        else
            last = i - index
        end
    end

    return last
end

start = map((x) -> parse(Int, x), split(readline(), ","))
first = play(start, 2020)
second = play(start, 30000000)
println("$(first) $(second)")
